target/linux/ipq806x/base-files/lib/upgrade/fortinet.sh

```diff
@@ -0,0 +1,30 @@
+. /lib/functions.sh
+
+# U-boot on the Fortinet FAP421E starts through the bootipq script
+# However, bootipq sets a custom bootarg that causes issues for us
+# This script patches that behavior, so that openwrt can boot normally
+# without requiring changes to the kernel
+
+fortinet_fap421e_patch_appsbl() {
```

Comment:
Isn't it possible to use `/chosen/bootargs` property to override ones passed by U-boot?
If not, care should be taken that the patching runs only during initial installation - I see the risk that after patching the U-boot initially, the error in line 23 might be falsely reported afterwards. I would consider verifying the checksum of U-boot partition, too.

Reply:
When I replaced the bootargs-append line with `bootargs = "console=ttyMSM0,9600n8";`, the device would stop at `Starting kernel ...` and reboot after a while.

Those values were copied from my device that boots OpenWrt properly from spiflash with a patched U-boot.

Ideally, I would have liked to use `/chosen/bootargs` too, but I am not sure if I am doing something incorrectly that is causing the bootargs to be ignored.

---

The script behavior has also been changed to check U-boot first, and only patch if necessary. The messages should also be clearer now.

A checksum was not used as it may be incompatible with other optional modifications e.g. removing their U-boot shell password. If a checksum is still preferred, I'll be glad to implement it.

Reply:
Ok, makes sense. I hope there was a safer solution than patching U-boot, but if this is necessary, then I'm perfectly fine with it.

Reply:
ttyMSM0 looks odd to me, are you sure this device is present after boot?

`ls /dev | grep tty`

Reply:
Yes, it is present. When sending `echo "hello" > /dev/ttyMSM0` over ssh, I can see the output on the serial console.
```
root@OpenWrt:/# ls /dev | grep tty
tty
ttyMSM0
ttyS0
ttyS1
```

Reply:
it may be possible that by defining `bootargs` with only the console info, that you're deleting something that's supposed to be there.

the beginning of kernel log shows the kernel command line, are there other things there before what's appended at the end?

Reply:
`console=ttyHSL0,9600n8` is present before the appended bootargs. 

I have tried `bootargs = "console=ttyHSL0,9600n8 console=ttyMSM0,9600n8"` on an unmodified U-boot, and it ends up with the same `Starting kernel ...`, and then restarting after a while.


```
root@OpenWrt:/# dmesg | grep command
[    0.000000] Kernel command line: console=ttyHSL0,9600n8 console=ttyMSM0,9600n8
[    0.000000] Bootloader command line not present
root@OpenWrt:/# cat /proc/cmdline
console=ttyHSL0,9600n8 console=ttyMSM0,9600n8
root@OpenWrt:/# cat /proc/device-tree/chosen/bootargs-append
 console=ttyMSM0,9600n8root@OpenWrt:/#
root@OpenWrt:/# cat /proc/device-tree/chosen/bootargs
console=ttyHSL0,9600n8root@OpenWrt:/#
```

Reply:
Have you tried ttyS0?

I'm starting to think that these other tty devices are not available in early parts of the boot, despite being usable after boot.

Also, what exactly is the bootarg that u-boot is adding that also causes issues? is it related to the `console` value or something else? Does it show under "bootloader command line"?

Reply:
or do you not even know the value of the bootarg that u-boot is passing, and only flipped a bit in order to stop that from executing?

Reply:
does removing this patch help?

target/linux/ipq806x/patches-6.1/0067-generic-Mangle-bootloader-s-kernel-arguments.patch

Reply:
nevermind, it seems like `bootargs` is being overridden regardless of entering a custom value in devicetree and it has nothing to do with that patch

in arch/arm/boot/compressed/atags_to_fdt.c
I would try filling the function `atags_to_fdt` with debug messages to see if that gets called and when, especially in the if...else where `merge_fdt_bootargs` is called, otherwise `setprop_string`

Reply:
The device starts from U-boot using `bootipq`, which loads the kernel from spiflash and appends `root=mtd:rootfs rootfstype=squashfs` to the bootargs. For example, the bootargs from the original Fortinet firmware looks like this:
```
FortiAP-421E # cat /proc/cmdline
console=ttyHSL0,9600n8 root=mtd:rootfs rootfstype=squashfs
```
The script `fortinet.sh` here removes those bootargs by replacing the first character with a null terminator. The default `console` parameter seems benign so I left it as-is. I was wondering if there was a better way to unset or override the `root` and `rootfstype` values.

Since the device boots using `bootipq`, values that were configured through `setenv` are ignored. Also, the env parameters are not persistent and will reset to baked-in defaults when restarted. I speculate that the U-boot was compiled with `CONFIG_ENV_IS_NOWHERE`.

Reply:
what about setting the values to null in bootargs

`root= rootfstype=`

Reply:
I tried setting empty values in both bootargs and bootargs-append. While in bootargs, the boot process behaved as though the entire property was absent and could not boot. In bootargs-append, I could at least see `root= rootfstype=` printed into the boot console. However the prior rootfs values seem to have taken precedence as it failed to boot when trying to identify the rootfs.

Out of curiosity, I tried these values in the devicetree:
```
chosen {
	bootargs-append = " console=ttyMSM0,9600n8 chosenbootargs=append";
	bootargs = "console=ttyMSM0,9600n8 chosenbootargs=set";
	append-rootblock = "nullparameter="; /* override the bootloader args */
};
```
With those values, this shows up on the bootlog. (the device still fails to find the rootfs)
```
[    0.000000] Kernel command line: console=ttyHSL0,9600n8 root=mtd:rootfs rootfstype=squashfs console=ttyMSM0,9600n8 chosenbootargs=append
[    0.000000] Bootloader command line not present
[    0.000000] Unknown kernel command line parameters "chosenbootargs=append", will be passed to user space.
```
Here, the `bootargs` property is completely ignored too. I also hoped that `append-rootblock` mangling would take effect and change `root=mtd:rootfs` into `nullparameter=mtd:rootfs` but that did not happen either.

I'm looking into the strategy of sprinkling `pr_notice()` logging in `atags_to_fdt.c`, but I am facing issues with this as I can't seem to add a reference to `_printk`. Is printk available at this part of the boot process?

Thanks for the suggestions and advice so far

Reply:
Turns out that the correct property was `bootargs-override`. Now the device boots fine without having to patch U-boot to remove the problematic bootargs. Thanks @Leo-PL and @mpratt14 !

Reply:
Oh that's interesting, I thought we were going to have to figure out how to make a property like that haha